File: labcpp/StaticLibrary1/vector.cpp
#include "vector.h"


Point::Point() {
    x = 0;
    y = 0;
}



double Point::getX() const {
    return x;
}

void Point::setX(double x) {
    this->x = x;
}

double Point::getY() const {
    return y;
}

void Point::setY(double y) {
    this->y = y;
}

bool Point::operator!() const
{ 
        return this->x == 0 && this->y == 0;
}

std::ostream& operator<<(std::ostream& os, const Point& point) {
    os << "x: " << point.getX() << ", y: " << point.getY() << "\n";
    return os;
}

Vector operator/(double number, Vector& vector)
{
    Vector result(vector);
    result.b.setX(result.b.getX() / number);
    result.b.setY(result.b.getY() / number);
    return result;
}

PolarCoordinates::PolarCoordinates(double r, double q) {
    this->r = r;
    this->q = q;
}

double PolarCoordinates::getR() const {
    return r;
}

void PolarCoordinates::setR(double r) {
    this->r = r;
}

double PolarCoordinates::getQ() const {
    return q;
}

void PolarCoordinates::setQ(double q) {
    this->q = q;
}


Vector Vector::operator/(double number) const
{
    Vector result(*this);
    result.b.setX(result.b.getX() / number);
    result.b.setY(result.b.getY() / number);
    return result;
}

Vector Vector::operator+(const Vector& other) const
{
    Vector result(*this);
    result.a.setX(result.a.getX() + other.a.getX());
    result.a.setY(result.a.getY() + other.a.getY());
    result.b.setX(result.b.getX() + other.b.getX());
    result.b.setY(result.b.getY() + other.b.getY());
    return result;
}

bool Vector::operator>(const Vector& other) const
{
    return this->Length() > other.Length();
}

bool Vector::operator<(const Vector& other) const
{
    return this->Length() < other.Length();
}

double Vector::Length() const
{
    return sqrt(pow(getCords().getX(), 2) + pow(getCords().getY(), 2));
}



bool Vector::operator!() const
{
    return !this->a || !this->b;
}

Vector Vector::operator--(int)
{
    Vector result(*this);
    b.setX(b.getX() - 1);
    b.setY(b.getY() - 1);
    return result;
}

Vector::Vector() {
    a = Point();
    b = Point();
}

Vector::Vector(Vector& vector)
{
    this->a = vector.getA();
    this->b = vector.getB();
}

Vector::Vector(const Vector& vector)
{
    this->a = vector.getA();
    this->b = vector.getB();
}

Vector::Vector(Point a, Point b) {
    this->a = a;
    this->b = b;
}

Point Vector::getA() const {
    return a;
}

void Vector::setA(Point a) {
    this->a = a;
}

Point Vector::getB() const {
    return b;
}

void Vector::setB(Point b) {
    this->b = b;
}

Point Vector::getCords() const {
    return Point(b.getX() - a.getX(), b.getY() - a.getY());
}

PolarCoordinates Vector::getPolarCoordinates() const {
    double r = sqrt(pow(getCords().getX(), 2) + pow(getCords().getY(), 2));
    double q = atan2(getCords().getY(), getCords().getX());
    return PolarCoordinates(r, q);
}

std::ostream& operator<<(std::ostream& os, const Vector& vector)
{
    os << "Point A: " << vector.getA() << "Point B: " << vector.getB() << "Coords: " << vector.getCords() << "Polar: r= "
         << vector.getPolarCoordinates().getR() << " q= " << vector.getPolarCoordinates().getQ() << "\n";
return os;
}

File: laboratorky/ClassLibrary1/Vector.cs
namespace ClassLibrary1;

public class Point
{
    private double _x;
    private double _y;

    public double X
    {
        get => _x;
        set => _x = value;
    }

    public double Y
    {
        get => _y;
        set => _y = value;
    }

    public static Point operator +(Point point1, Point point2)
    {
        return new Point(point1.X + point2.X, point1.Y + point2.Y);
    }
    public static bool operator ! (Point point)
    {
        return point.X == 0 && point.Y == 0;
    }
    public Point()
    {
        this.X = 0;
        this.Y = 0;
    }
    public Point(Point point)
    {
        this.X = point.X;
        this.Y = point.Y;
    }

    public Point(double x, double y)
    {
        this.X = x;
        this.Y = y;
    }

    public override string ToString()
    {
        return $"x: {X},y: {Y}\n";
    }
}

public class PolarCoorditates
{
    private double _r;
    private double _q;

    public double R
    {
        get => _r;
        set => _r = value;
    }

    public double Q
    {
        get => _q;
        set => _q = value;
    }

    public PolarCoorditates(double r, double q)
    {
        this.R = r;
        this.Q = q;
    }
}

public class Vector
{
    public Point Cords => new Point(B.X - A.X, B.Y - A.Y);

    public double Length => Math.Sqrt(Math.Pow(Cords.X, 2) + Math.Pow(Cords.Y, 2));
    public PolarCoorditates Polar => PolarCoorditates();

    private Point _a;
    private Point _b;
    public static Vector operator /(Vector vector, double number)
    {
        return new Vector(vector.A, new Point(vector.B.X / number, vector.B.Y / number));
    }
    public  static Vector operator /(double number, Vector vector)
    {
        return new Vector(vector.A, new Point(vector.B.X / number, vector.B.Y / number));
    }
    public static Vector operator +(Vector vector1, Vector vector2)
    {
        return new Vector(new Point(vector1.A + vector2.A), new Point(vector1.B + vector2.B));
    }
    public static bool operator > (Vector vector1, Vector vector2)
    {
        return vector1.Length > vector2.Length;
    }
    public static bool operator < (Vector vector1, Vector vector2)
    {
        return vector1.Length < vector2.Length;
    }
    public static Vector operator -- (Vector vector)
    {
        Vector vector1 = new Vector(vector);
        vector1.B.X--;
        vector1.B.Y--;
        return vector1;
    }
    public static bool operator ! (Vector vector)
    {
        return !vector.A || !vector.B; 
    }
    public Point A
    {
        get => _a;
        set => _a = value;
    }

    public Point B
    {
        get => _b;
        set => _b = value;
    }


    public PolarCoorditates PolarCoorditates()
    {
        var r = Math.Sqrt((Math.Pow(Cords.X, 2)) + Math.Pow(Cords.Y, 2));
        var q = Math.Atan2(Cords.Y, Cords.X);
        return new PolarCoorditates(r, q);
    }

    public Vector(Vector vector)
    {
        this.A = new Point(vector.A);
        this.B = new Point(vector.B);
    }
    public Vector()
    {
        this.A = new Point();
        this.B = new Point();
    }

    public Vector(Point a, Point b)
    {
        this.A = a;
        this.B = b;
    }
    

    public override string ToString()
    {
        return
            $"Point A: {A.ToString()}Point B: {B.ToString()}Coords:{Cords.ToString()}Polar: r= {Polar.R}, q= {Polar.Q}";
    }
}